### ISSUE TITLE
Lexer refactor

### ISSUE THREAD

**Issue by znichola** [sophiakoulen/mathematical-expression-parser#8]
The more I think about it, the better it seems to have a list of tokens. It can avoid a whole bunch of lines that call free, or checking the token is valid etc ... which helps to reduce code length and for the `^` which is right associative it's easier to look ahead without worrying about reading the string and keeping track of position and checking what needs to be freed etc.

_Originally posted by @znichola in https://github.com/sophiakoulen/mathematical-expression-parser/issues/5#issuecomment-1372308654_

**Comment by znichola** [comment]
https://github.com/znichola/mathematical-expression-parser/commit/2b4588ea0d8ea9877837a7d51391f4fad10f648a here I've added the lexer, with room I think for expansion to include words and constants.
The list of tokens is stored as a doubly linked list reusing the `t_tree` stuct
I've had to add some token types to the `e_token_type` but am a bit un sure on how to continue with this.
I sort the token into 3 basic types:
* value  `FLOAT`
* symbol `"-+*/^()"` one of these
* end `'\0'` for the end of the line
* invalid - for any other charater that's read

Next step is to refactor the parser!
